src/project.coffee

```diff
@@ -96,11 +110,45 @@ class Project extends Model
 
   # Public: Get an {Array} of {GitRepository}s associated with the project's
   # directories.
+  #
+  # This method will be removed in 2.0 because it does synchronous I/O.
+  # Prefer the following, which evaluates to a Promise that resolves to an
+  # {Array} of {Repository} objects:
+  # ```
+  # project.getDirectories().map(
+  #     project.repositoryForDirectory.bind(project))
+  # ```
   getRepositories: -> _.compact([@repo])
   getRepo: ->
     Grim.deprecate("Use ::getRepositories instead")
     @repo
 
+  # Public: Returns a {Promise} that resolves with either:
+  # * {Repository} if a repository can be created for the given directory
+  # * `null` if no repository can be created for the given directory.
+  #
+  # * `directory` {Directory} for which to get a {Repository}.
+  repositoryForDirectory: (directory) ->
+    path = directory.getRealPathSync()
+    promise = @directoryToRepositoryMap.get(path)
+    unless promise
+      promises = @repositoryProviders.map(
+          (provider) -> provider.repositoryForDirectory directory)
+      promise = Promise.all(promises).then((repositories) =>
```

Comment:
I'm pretty sure the following formatting would work, which I think could be more clear:

```coffee
promise = Promise.all(promises).then
  (repositories) =>
    # now the function is indented and doesn't need the parens, which looks weird on line 148 with `repo)`
```